Adding functional tests for fusion_hap, fusion_hw, fusion_nig.

andreepyro opened the pull request:
##### SUMMARY
Add functional tests.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
- tests

##### ADDITIONAL INFORMATION
Adding functional tests for fusion_hap, fusion_hw, fusion_nig.

jankodera-ps commented on plugins/modules/fusion_hap.py:
```diff
@@ -135,14 +135,11 @@
+    for host in hosts:
+        if host.iqn == module.params["iqn"] and host.name != module.params["name"]:
             module.fail_json(
                 msg="Supplied IQN {0} already used by host access polivy {1}".format(
```
I know its not your code, but there is a typo here and your PR is in ideal position to fix it 😁

andreepyro replied:
Sure :)

jankodera-ps commented on tests/functional/test_fusion_hap.py:
```diff
@@ -0,0 +1,1024 @@
+        "nqn": "string_value",
+        "iqn": "iqn.2023-05.com.purestorage:420qp2c0699",
+        "wwns": ["wwn1", "wwn2"],
+        "host_password": "u^nv2320XnYQ",
```
I didn't notice before that Ansible HAP has these. We should remove them, they are not supported and will just cause confusion. Lets do it in a separate task.

andreepyro replied:
Agree.

jankodera-ps commented on tests/functional/test_fusion_hap.py:
```diff
@@ -0,0 +1,1024 @@
+        (HTTPError, HTTPError),
+    ],
+)
+def test_az_update_exception(
```
These stubs should probably be named `test_hap_...`. However to tell the truth I would not add these stubs at all. By doing so you're creating an expectation which does not necessarily have to be met, since it's out of our control and can confuse the users.

jankodera-ps replied:
If you want to have a reminder, create tasks for it rather than leaving such stubs. Tasks are actionable.

andreepyro replied:
You're right. I will remove it also from the `test_fusion_az.py` file.

jankodera-ps commented on tests/functional/test_fusion_hw.py:
```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*-
```
It does not make sense to add tests for a deprecated module. Deprecation is only a mechanism to not break existing functionality before users had chance to migrate off that module, the code will not be further developed and will be removed, so spending additional work on it is wasted in my eyes.

jankodera-ps replied:
Additionally the extra code is something you have to take in consideration in any further tasks, this thankfully does not impact Python as it is a fully dynamic and does not have compile time checks which would require fixes.

andreepyro replied:
> It does not make sense to add tests for a deprecated module. Deprecation is only a mechanism to not break existing functionality before users had chance to migrate off that module, the code will not be further developed and will be removed, so spending additional work on it is wasted in my eyes.

I think that while this module is still part of the collection it should be tested. Also, it didn't require much time from my side.

> Additionally the extra code is something you have to take in consideration in any further tasks, this thankfully does not impact Python that much as it is a fully dynamic and does not have compile time checks which would require fixes, but the tests still get run and have to pass.

If it was integration test or e2e test, yes. This runs almost instantly.

jankodera-ps replied:
It is not about running time but rather about maintenance burden, the effort is usually better spent elsewhere. There is always other work that needs to be done and perfectionism rarely pays off :)

andreepyro replied:
Hmm, you're right. The maintenance doesn't worth it.

jankodera-ps commented on plugins/modules/fusion_nig.py:
```diff
@@ -117,9 +117,6 @@
     setup_fusion,
 )
-from ansible_collections.purestorage.fusion.plugins.module_utils.getters import (
-    get_az,
```
I wonder if this should go the other way, i.e. whether get_az and get_region checks should be added to make this module consistent with the other ones

jankodera-ps replied:
Nevermind, the general direction is to remove them.

andreepyro replied:
I don't think we should call any other APIs in the modules (except where it is required). Not only it adds additional logic to the module, but it also makes more http requests that wouldn't be needed. Also, if you insert incorrect `availability_zone`, you'll get roughly the same error (the API will tell you).

andreepyro commented:
@jankodera-ps sorry for the force push. What I did:
- fix the typo in fusion_hap.py
- removed stub tests from test_fusion_hap and test_fusion_az
- removed testing of deprecated parameters in test_fusion_hap

andreepyro commented:
@sdodsley please review

andreepyro commented:
Rebased onto the latest master.
